File: bindings/kotlin/src/nativeMain/kotlin/org/maplibre/nativeffi/map/CustomMvtVectorSourceState.kt
package org.maplibre.nativeffi.map

import kotlinx.cinterop.COpaquePointer
import kotlinx.cinterop.CPointer
import kotlinx.cinterop.CValue
import kotlinx.cinterop.ExperimentalForeignApi
import kotlinx.cinterop.StableRef
import kotlinx.cinterop.alloc
import kotlinx.cinterop.asStableRef
import kotlinx.cinterop.nativeHeap
import kotlinx.cinterop.ptr
import kotlinx.cinterop.staticCFunction
import kotlinx.cinterop.useContents
import org.maplibre.nativeffi.internal.c.MLN_CUSTOM_MVT_VECTOR_SOURCE_OPTION_MAX_ZOOM
import org.maplibre.nativeffi.internal.c.MLN_CUSTOM_MVT_VECTOR_SOURCE_OPTION_MIN_ZOOM
import org.maplibre.nativeffi.internal.c.mln_canonical_tile_id
import org.maplibre.nativeffi.internal.c.mln_custom_mvt_vector_source_options
import org.maplibre.nativeffi.internal.c.mln_custom_mvt_vector_source_options_default
import org.maplibre.nativeffi.internal.callback.CallbackGate
import org.maplibre.nativeffi.internal.struct.StyleStructs
import org.maplibre.nativeffi.style.CustomMvtVectorSourceOptions

/**
 * Owns map/style-scoped custom MVT vector source callback state.
 *
 * [onReleased] runs on the map owner thread when native stops referencing this state, which is what
 * drops it from its map's registry and closes it.
 */
@OptIn(ExperimentalForeignApi::class)
internal class CustomMvtVectorSourceState(
  private val options: CustomMvtVectorSourceOptions,
  private val onReleased: () -> Unit,
) : AutoCloseable {
  private val selfRef = StableRef.create(this)
  private val descriptor = nativeHeap.alloc<mln_custom_mvt_vector_source_options>()
  private val gate = CallbackGate("custom MVT vector callbacks") { closeNative() }

  init {
    mln_custom_mvt_vector_source_options_default().place(descriptor.ptr)
    descriptor.fetch_tile = staticCFunction(::customMvtVectorFetchTile)
    descriptor.cancel_tile = staticCFunction(::customMvtVectorCancelTile)
    descriptor.release_user_data = staticCFunction(::customMvtVectorReleaseUserData)
    descriptor.user_data = selfRef.asCPointer()
    writeFields(descriptor)
  }

  fun descriptor(): CPointer<mln_custom_mvt_vector_source_options> = descriptor.ptr

  internal fun fetch(tileId: CValue<mln_canonical_tile_id>) {
    val lease = gate.enter() ?: return
    try {
      options.callback.fetchTile(tileId.useContents { StyleStructs.canonicalTileId(this) })
    } catch (_: Throwable) {
      // Native callbacks must not unwind through the C ABI.
    } finally {
      lease.close()
    }
  }

  /** Runs when native stops referencing this state, on the map owner thread. */
  internal fun released() {
    onReleased()
  }

  internal fun cancel(tileId: CValue<mln_canonical_tile_id>) {
    val lease = gate.enter() ?: return
    try {
      options.callback.cancelTile(tileId.useContents { StyleStructs.canonicalTileId(this) })
    } catch (_: Throwable) {
      // Native callbacks must not unwind through the C ABI.
    } finally {
      lease.close()
    }
  }

  private fun writeFields(native: mln_custom_mvt_vector_source_options) {
    native.fields = 0U
    options.minZoom?.let {
      native.fields = native.fields or MLN_CUSTOM_MVT_VECTOR_SOURCE_OPTION_MIN_ZOOM
      native.min_zoom = it
    }
    options.maxZoom?.let {
      native.fields = native.fields or MLN_CUSTOM_MVT_VECTOR_SOURCE_OPTION_MAX_ZOOM
      native.max_zoom = it
    }
  }

  override fun close() = gate.close()

  private fun closeNative() {
    selfRef.dispose()
    nativeHeap.free(descriptor.rawPtr)
  }
}

@Suppress("UNUSED_PARAMETER")
@OptIn(ExperimentalForeignApi::class)
private fun customMvtVectorFetchTile(
  userData: COpaquePointer?,
  tileId: CValue<mln_canonical_tile_id>,
) {
  userData?.asStableRef<CustomMvtVectorSourceState>()?.get()?.fetch(tileId)
}

@Suppress("UNUSED_PARAMETER")
@OptIn(ExperimentalForeignApi::class)
private fun customMvtVectorCancelTile(
  userData: COpaquePointer?,
  tileId: CValue<mln_canonical_tile_id>,
) {
  userData?.asStableRef<CustomMvtVectorSourceState>()?.get()?.cancel(tileId)
}

@OptIn(ExperimentalForeignApi::class)
private fun customMvtVectorReleaseUserData(userData: COpaquePointer?) {
  try {
    userData?.asStableRef<CustomMvtVectorSourceState>()?.get()?.released()
  } catch (_: Throwable) {
    // Native callbacks must not unwind through the C ABI.
  }
}

File: bindings/dotnet/src/Maplibre.NativeFfi/Internal/Callback/CustomMvtVectorSourceState.cs
using System.Runtime.InteropServices;
using Maplibre.NativeFfi.Geo;
using Maplibre.NativeFfi.Internal.C;
using Maplibre.NativeFfi.Internal.Loader;
using Maplibre.NativeFfi.Style;

namespace Maplibre.NativeFfi.Internal.Callback;

internal sealed unsafe class CustomMvtVectorSourceState : IDisposable
{
    private readonly Lock gate = new();
    private readonly CustomMvtVectorSourceOptions options;
    private readonly CustomMvtVectorSourceCallback fetchTile;
    private readonly CustomMvtVectorSourceCallback? cancelTile;
    private GCHandle handle;
    private bool closed;
    private bool handleFreed;
    private int activeCallbacks;

    internal CustomMvtVectorSourceState(CustomMvtVectorSourceOptions options)
    {
        this.options = options ?? throw new ArgumentNullException(nameof(options));
        fetchTile =
            options.FetchTile
            ?? throw new ArgumentException(
                "Custom MVT vector source FetchTile callback is required.",
                nameof(options)
            );
        cancelTile = options.CancelTile;
        handle = GCHandle.Alloc(this);
    }

    internal mln_custom_mvt_vector_source_options Descriptor
    {
        get
        {
            NativeLibraryLoader.EnsureLoaded();
            var descriptor = NativeMethods.mln_custom_mvt_vector_source_options_default();
            descriptor.fetch_tile = &FetchTile;
            descriptor.cancel_tile = &CancelTile;
            descriptor.release_user_data = &ReleaseUserData;
            descriptor.user_data = (void*)GCHandle.ToIntPtr(handle);
            if (options.MinimumZoom is { } minimumZoom)
            {
                descriptor.fields |= (uint)
                    mln_custom_mvt_vector_source_option_field.MLN_CUSTOM_MVT_VECTOR_SOURCE_OPTION_MIN_ZOOM;
                descriptor.min_zoom = minimumZoom;
            }
            if (options.MaximumZoom is { } maximumZoom)
            {
                descriptor.fields |= (uint)
                    mln_custom_mvt_vector_source_option_field.MLN_CUSTOM_MVT_VECTOR_SOURCE_OPTION_MAX_ZOOM;
                descriptor.max_zoom = maximumZoom;
            }
            return descriptor;
        }
    }

    internal bool IsHandleAllocatedForTest => handle.IsAllocated;

    internal void FetchForTest(CanonicalTileId tileId) => InvokeFetch(tileId);

    internal void CancelForTest(CanonicalTileId tileId) => InvokeCancel(tileId);

    [UnmanagedCallersOnly(CallConvs = [typeof(System.Runtime.CompilerServices.CallConvCdecl)])]
    private static void FetchTile(void* userData, mln_canonical_tile_id tileId)
    {
        var state = FromUserData(userData);
        state?.InvokeFetch(FromNative(tileId));
    }

    [UnmanagedCallersOnly(CallConvs = [typeof(System.Runtime.CompilerServices.CallConvCdecl)])]
    private static void CancelTile(void* userData, mln_canonical_tile_id tileId)
    {
        var state = FromUserData(userData);
        state?.InvokeCancel(FromNative(tileId));
    }

    // The C API invokes this once, on the runtime's native scheduler thread, after MapLibre stops referencing this
    // state, so the stubs go with the source whether it was removed, dropped by a style load, or
    // retired with the map.
    [UnmanagedCallersOnly(CallConvs = [typeof(System.Runtime.CompilerServices.CallConvCdecl)])]
    private static void ReleaseUserData(void* userData)
    {
        var state = FromUserData(userData);
        state?.Dispose();
    }

    private static CustomMvtVectorSourceState? FromUserData(void* userData)
    {
        try
        {
            return (CustomMvtVectorSourceState?)GCHandle.FromIntPtr((nint)userData).Target;
        }
        catch
        {
            return null;
        }
    }

    private static CanonicalTileId FromNative(mln_canonical_tile_id tileId) =>
        new(tileId.z, tileId.x, tileId.y);

    private void InvokeFetch(CanonicalTileId tileId)
    {
        if (!EnterCallback())
        {
            return;
        }

        try
        {
            fetchTile.Invoke(tileId);
        }
        catch
        {
            // Native callbacks must not unwind through the C ABI.
        }
        finally
        {
            ExitCallback();
        }
    }

    private void InvokeCancel(CanonicalTileId tileId)
    {
        if (!EnterCallback())
        {
            return;
        }

        try
        {
            cancelTile?.Invoke(tileId);
        }
        catch
        {
            // Native callbacks must not unwind through the C ABI.
        }
        finally
        {
            ExitCallback();
        }
    }

    private bool EnterCallback()
    {
        lock (gate)
        {
            if (closed)
            {
                return false;
            }

            activeCallbacks++;
            return true;
        }
    }

    private void ExitCallback()
    {
        lock (gate)
        {
            activeCallbacks--;
            FreeHandleIfReadyLocked();
        }
    }

    public void Dispose()
    {
        lock (gate)
        {
            closed = true;
            FreeHandleIfReadyLocked();
        }
    }

    private void FreeHandleIfReadyLocked()
    {
        if (!closed || activeCallbacks != 0 || handleFreed || !handle.IsAllocated)
        {
            return;
        }

        handle.Free();
        handleFreed = true;
    }
}
